### PR TITLE
Enable batch matmul for result sizes > 2**32 the tensor can be split along batch axis

### LINKED ISSUE #131865: MPS backend breaking on llama 3.1 8B on Macbook M3
### 🐛 Describe the bug

python code being run under the hood:

```python

_TORCH_DEVICE = "mps" if torch.backends.mps.is_available() else "cpu"
print(f"Using torch device: {_TORCH_DEVICE}")


@functools.cache
def _build_llama_pipeline():
    pipeline = transformers.pipeline(
        "text-generation",
        model="meta-llama/Meta-Llama-3.1-8B-Instruct",
        device=_TORCH_DEVICE,
    )
    return pipeline

...
messages = [
        {"role": "system", "content": prompt},
    ]
pipe = _build_llama_pipeline()
result = pipe(messages, max_new_tokens=200)
```

I am getting the following error:
```bash
/AppleInternal/Library/BuildRoots/d615290d-668b-11ee-9734-0697ca55970a/Library/Caches/com.apple.xbs/Sources/MetalPerformanceShaders/MPSCore/Types/MPSNDArray.mm:761: failed assertion `[MPSNDArray initWithDevice:descriptor:] Error: total bytes of NDArray > 2**32'
```

This is a device issue, the same code runs fine on the cpu. I am installing torch via `poetry add 'transformers[torch]'`

Related issues I've found on this topic (without a clear solution):
https://github.com/pytorch/pytorch/issues/78247
https://github.com/pytorch/pytorch/issues/78492

### Versions

Collecting environment information...
PyTorch version: 2.4.0
Is debug build: False
CUDA used to build PyTorch: None
ROCM used to build PyTorch: N/A

OS: macOS 14.1 (arm64)
GCC version: Could not collect
Clang version: 15.0.0 (clang-1500.1.0.2.5)
CMake version: Could not collect
Libc version: N/A

Python version: 3.12.4 (main, Jul 13 2024, 01:18:59) [Clang 15.0.0 (clang-1500.1.0.2.5)] (64-bit runtime)
Python platform: macOS-14.1-arm64-arm-64bit
Is CUDA available: False
CUDA runtime version: No CUDA
CUDA_MODULE_LOADING set to: N/A
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Apple M3 Max

Versions of relevant libraries:
[pip3] mypy==1.11.0
[pip3] mypy-extensions==1.0.0
[pip3] numpy==2.0.1
[pip3] torch==2.4.0
[conda] Could not collect

cc @ezyang @gchanan @zou3519 @kadeng @msaroufim @kulinseth @albanD @malfet @DenisVieriu97 @jhavukainen

### PR DESCRIPTION
**Pull request by jhavukainen**
Fixes #131865. Addresses the issue seen when running llama v3.1 8B parameter model on MPS backend where the batch matmul output size can go over the 32-bit indexing limit of MPS tensors, causing an assert.

Test case to reproduce the issue with the dimensions encountered in llama v3.1 and verify this fix works around it:

```
import torch
device='mps'
a = torch.randn([32, 20064, 128], dtype=torch.float32,device=device)
b = torch.randn([32, 128, 20064], dtype=torch.float32, device=device)
res = torch.bmm(a, b)
```

Notably the current change only works as long as the individual output matrix in the bmm does not exceed the number of elements 2**32. This lets us split up the computation along the batch axis to avoid going over the limit. 

Added a TORCH_CHECK to raise an error if the individual matrix dimensions are too large to handle for this op until a more general workaround tiling the matmuls is available.

### CODE REVIEW

**Comment by jhavukainen** [comment]
Need some guidance on how do we want to test this: 

In order to reach the original failure a 96GB configuration is required. With a 64GB config we will hit another error when trying to allocate the initial buffers: `RuntimeError: Invalid buffer size: 47.99 GB`. This is due to the max memory fraction pytorch is allowed.

In order to validate that the MPS implementation matches the CPU implementation we'll need a 128GB config. This is because the output tensors in the test case that fails are around ~50GB each leading to my 96GB config killing the process when asserting that the CPU and MPS outputs are the same. This can however be worked around by reducing the tensors a bit, batch_size = 11 is just enough to take the new code path. That seems to be able to run all the way including the result validation on a 96GB config.

@kulinseth @malfet Do we want guarantee our mac configurations now and going forwards are at least 96GB ones? Or should we try to work around this in some other way? The first I can think of would be to add a skip / xFail decorator if we detect that the test is running on a config with less than 96GB. I'm thinking it might have to be a skip since if the python process hogs too much memory it looks like the OS just kills it off instead of the test just failing due to some assert for not being able to allocate more memory.

**Comment by malfet** [comment]
@jhavukainen I'm a bit confused about the repro case, because llama-3.1 should never run in full 32-bit mode, as it will be extremely slow and at least for 8B model would have zero advantage over bfloat16 variant that uses less memory

**Comment by jhavukainen** [comment]
> @jhavukainen I'm a bit confused about the repro case, because llama-3.1 should never run in full 32-bit mode, as it will be extremely slow and at least for 8B model would have zero advantage over bfloat16 variant that has a 2x smaller memory footprint)

Well yes that is a bit confusing. But it comes directly from the developer's provided repro script where their llama v3.1 pipeline gives
```
print(pipeline.torch_dtype)
# torch.float32
```
and leads to the op that fails being also of type f32 `aten::bmm_out_mps_impl:f32[32,20064,128]:f32[32,128,20064]`.

But now that you mentioned it I tested the behavior when running with the half data types and there is clearly some other issue we need to tackle there since the MPSMatrixMultplication is doing some op underneath in those cases, leading to the same error in a slightly different context. And for bfloat16 specifically it raises an assert for unsupported data type so I'll need to consult the MPS team on workaround for that.

While I wait for their feedback to see what should be done I'll add a check for the data type to raise the unsupported error for the half types for now. Since fp32 is the specific case the developer asked for this PR would still unblock them.

**Comment by jhavukainen** [comment]
Updated the PR to work also on float16 and bfloat16. With this the LLaMA3.1 works on all the supported float data types. Now I'll just need to figure a precompiler directive to hide it from pre-MacOS15 machines since the compiler will refuse the `MPSNDArrayDescriptor.preferPackedRows` on those machines.

**Comment by jhavukainen** [comment]
@malfet @kulinseth I ended up adding the tests for fp16 and bf16 since those have a smaller memory footprint. The tests will  still take a quite a bit more time than many other unittests we have, around 2 minutes per data type on a powerful machine. This is mostly down to how long the CPU takes computing the reference value and doing the comparison. Let me know if this is a concern and if there are other ideas on what would be a better test.

Other than that this PR should be ready to review.

**Comment by malfet** [comment]
> Other than that this PR should be ready to review.
@jhavukainen isn't it just failed on MacOS 15, see https://github.com/pytorch/pytorch/actions/runs/10620567384/job/29441019041

File: aten/src/ATen/native/mps/operations/LinearAlgebra.mm
Comment (malfet) [inline aten/src/ATen/native/mps/operations/LinearAlgebra.mm:513]: This would never gets compiled, would it? (As our builders are still target MacOS-13 as min supported OS)
You'll need to use `if (is_macos_13_or_newer(MacOSVersion::MACOS_VER_15_0_PLUS)))

File: aten/src/ATen/native/mps/operations/LinearAlgebra.mm
Comment (malfet) [inline aten/src/ATen/native/mps/operations/LinearAlgebra.mm:547]: Just curious, what's wrong with autos? 
```suggestion
      auto matmul = [[MPSNDArrayMatrixMultiplication alloc] initWithDevice:device
```

File: aten/src/ATen/native/mps/operations/LinearAlgebra.mm
Comment (jhavukainen) [reply aten/src/ATen/native/mps/operations/LinearAlgebra.mm:513]: I see. That's a bummer. Then we'll need to do this in a roundabout way of adding the API definitions that come up in MacOS15 to the code somewhere. Else the runtime check won't help as the compiler kills the build beforehand.

**Comment by jhavukainen** [comment]
MacOS15 now passing.

File: aten/src/ATen/native/mps/operations/LinearAlgebra.mm
Comment (jhavukainen) [reply aten/src/ATen/native/mps/operations/LinearAlgebra.mm]: Done.

File: test/test_mps.py
Comment (malfet) [inline test/test_mps.py:1944]:
```suggestion
    @parametrize("dtype", [torch.float16, torch.bfloat16])
    def test_large_bmm_bf16(self, dtype):
        batch1 = torch.randn(11, 20064, 128, dtype=dtype, device='mps')
        batch2 = torch.randn(11, 128, 20064, dtype=dtype, device='mps')
        output_cpu = torch.bmm(batch1.cpu(), batch2.cpu())
        output_mps = torch.bmm(batch1, batch2)

        # Using the low precision comparison for FP16
        tol = 1e-2 if dtype == torch.float16 else None
        self.assertEqual(output_cpu, output_mps, atol=tol, rtol=tol)
        self.assertEqual(output_cpu.size(), output_mps.size())

```

**Review by malfet — APPROVED** [review]

**Pull request closed without merging**
